src/libsystemd/sd-bus/bus-socket.c

```diff
@@ -1037,8 +1037,9 @@ int bus_socket_write_message(sd_bus *bus, sd_bus_message *m, size_t *idx) {
                 if (m->n_fds > 0 && *idx == 0) {
                         struct cmsghdr *control;
 
-                        mh.msg_control = control = alloca(CMSG_SPACE(sizeof(int) * m->n_fds));
-                        mh.msg_controllen = control->cmsg_len = CMSG_LEN(sizeof(int) * m->n_fds);
+                        mh.msg_controllen = CMSG_SPACE(sizeof(int) * m->n_fds);
+                        mh.msg_control = control = alloca(mh.msg_controllen);
+                        control->cmsg_len = CMSG_LEN(sizeof(int) * m->n_fds);
```

Comment:
cmsg(3) says
> When initializing a buffer that will contain a series of  cmsghdr  structures (e.g., to be sent with sendmsg(2)), that buffer should first be zero-initialized to ensure the correct operation of CMSG_NXTHDR().

Confusingly, the example in the man page does not perform said initialization.
Nevertheless, we usually do zero initialization. Here it is not done, which is at least inconsistent.

Also, I think we are supposed to use CMSG_FIRSTHDR. Shouldn't it also be used here?

Reply:
true, and true